File: docker/dotnet/CreateYml.cs
using System;
using System.Xml;
using System;
using YamlDotNet.RepresentationModel;
using YamlDotNet.Serialization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using static DotNetPreAdapter.Datastructure;
using System.Collections;
using YamlDotNet.Core.Tokens;
using System.IO;

// dotnet add package YamlDotNet --version 12.2.0
namespace DotNetPreAdapter
{
    public class CreateYml
    {
        public void Create(Datastructure datastructure, string toDir)
        {

            var subDatastructureYmlNode = new YamlSequenceNode();
            foreach (string classname in datastructure.get_classname_list())
            {
                SubDataStructure? subDataStructure = datastructure.get_datastructures_from_class_name(classname);
                if (subDataStructure != null)
                {
                    var fromImports = new YamlSequenceNode();
                    foreach (var (imported_class_name, namespace_path) in subDataStructure.get_from_imports())
                    {
                        fromImports.Add(
                            new YamlMappingNode(
                                new YamlScalarNode("imported_class_name"),
                                    new YamlScalarNode(imported_class_name),
                                new YamlScalarNode("namespace_path"),
                                    new YamlScalarNode(namespace_path)
                                )
                            );
                    }

                    var innerClasses = new YamlSequenceNode();
                    foreach (string class_name in subDataStructure.get_inner_class_name())
                    {
                        innerClasses.Add(new YamlScalarNode(class_name));
                    }
                    var anonymousCalls = new YamlSequenceNode();
                    foreach (string anonymousCall in subDataStructure.getAnonymousInvocation())
                    {
                        anonymousCalls.Add(
                                    new YamlScalarNode(anonymousCall)
                            );
                    }
                    var baseClasses = new YamlSequenceNode();
                    foreach (string baseClass in subDataStructure.get_base_classes())
                    {
                        baseClasses.Add(
                                    new YamlScalarNode(baseClass)
                            );
                    }
                    var statics = new YamlSequenceNode();
                    foreach (Static staticTypes in subDataStructure.get_static_fields())
                    {
                        statics.Add(
                            new YamlMappingNode(
                                new YamlScalarNode("static_name"),
                                    new YamlScalarNode(staticTypes.StaticName),
                                new YamlScalarNode("static_type"),
                                    new YamlScalarNode(staticTypes.StaticType)
                                )
                            );
                    }
                    var variables = new YamlSequenceNode();
                    foreach (Variable variable in subDataStructure.get_variable_fields())
                    {
                        variables.Add(
                                new YamlMappingNode(
                                    new YamlScalarNode("variable_name"),
                                        new YamlScalarNode(variable.VariableName),
                                    new YamlScalarNode("variable_type"),
                                        new YamlScalarNode(variable.VariableType),
                                    new YamlScalarNode("is_member"),
                                        new YamlScalarNode(variable.IsMember ? "True" : "False")
                                )
                        );
                    }

                    var methods = new YamlSequenceNode();
                    foreach (Method method in subDataStructure.get_method_fields())
                    {
                        var parameters = new YamlSequenceNode();
                        foreach(var parameter in method.Parameters)
                        {
                            parameters.Add(new YamlMappingNode(
                                    new YamlScalarNode("parameter_name"),
                                        new YamlScalarNode(parameter.Parameter),
                                    new YamlScalarNode("parameter_type"),
                                        new YamlScalarNode(parameter.UserType)));
                        }
                        methods.Add(
                                new YamlMappingNode(
                                    new YamlScalarNode("method_name"),
                                    new YamlScalarNode(method.MethodName),
                                    new YamlScalarNode("parameters"),
                                    parameters,
                                    new YamlScalarNode("is_private"),
                                    new YamlScalarNode(method.IsPrivate ? "True" : "False")
                                    )
                                );
                    }


                    subDatastructureYmlNode.Add(new YamlMappingNode(
                        new YamlScalarNode("sub_datastructure"),
                        new YamlMappingNode(
                            new YamlScalarNode("fqdn_class_name"),
                                new YamlScalarNode(subDataStructure.get_fqdn_class_name()),

                            new YamlScalarNode("filename"),
                                new YamlScalarNode(subDataStructure.get_filename()),
                            new YamlScalarNode("from_imports"),
                                fromImports,
                            new YamlScalarNode("anonymous_calls"),
                                anonymousCalls,
                            new YamlScalarNode("inner_classes"),
                                innerClasses,
                            new YamlScalarNode("variables"),
                                variables,
                            new YamlScalarNode("is_abstract"),
                                new YamlScalarNode(subDataStructure.is_abstract() ? "True" : "False"),
                            new YamlScalarNode("is_interface"),
                                new YamlScalarNode(subDataStructure.IsInterface() ? "True" : "False"),

                            new YamlScalarNode("base_classes"),
                                baseClasses,

                            new YamlScalarNode("statics"),
                                statics,
                            new YamlScalarNode("methods"),
                                methods,

                            new YamlScalarNode("filemodule"),
                                new YamlScalarNode(subDataStructure.get_filemodule())

                                )
                            )
                        );
                }
            }

            var stream = new YamlStream(
                new YamlDocument(subDatastructureYmlNode
                  
                    )
                );
            using (TextWriter writer = File.CreateText(toDir + "/output.yaml"))
            {
                stream.Save(writer, false);
            }
        }
    }
}


                

File: java-adapter/src/main/java/preadapter/infrastructure/CreateYML.java
package preadapter.infrastructure;

import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.error.Mark;
import org.yaml.snakeyaml.nodes.*;
import preadapter.Logger;
import preadapter.domain.Datastructure;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.*;

public class CreateYML {

    private final DumperOptions.FlowStyle dummperOptions = DumperOptions.FlowStyle.AUTO;
    private final Mark startMark = null;
    private final Mark endMark = null;

    private boolean appendTupleNodeValueSequenceNode(MappingNode mappingNode, String key, SequenceNode value) {

        return mappingNode.getValue().add( new NodeTuple(
                new ScalarNode(Tag.STR, key, startMark, endMark, DumperOptions.ScalarStyle.PLAIN),
                value
        ));
    }

    private boolean appendTupleNodeValueString(MappingNode mappingNode, String key, String value) {

        return mappingNode.getValue().add( new NodeTuple(
                new ScalarNode(Tag.STR, key, startMark, endMark, DumperOptions.ScalarStyle.PLAIN),
                new ScalarNode(Tag.STR, value, startMark, endMark, DumperOptions.ScalarStyle.PLAIN)
        ));
    }
    private boolean appendTupleNodeValueStringToSequenceNode(SequenceNode sequenceNode, String key, String value) {
        var mappingNode = new MappingNode(
                Tag.MAP, new ArrayList<>(), dummperOptions);
        mappingNode.getValue().add( new NodeTuple(
                new ScalarNode(Tag.STR, key, startMark, endMark, DumperOptions.ScalarStyle.PLAIN),
                new ScalarNode(Tag.STR, value, startMark, endMark, DumperOptions.ScalarStyle.PLAIN)
        ));
        return sequenceNode.getValue().add(mappingNode);
    }

    private void appendSequence(MappingNode mappingNode, List<String> list, String nodeName) {

        SequenceNode subSequenceNode = new SequenceNode(
                Tag.SEQ, new ArrayList<>(), dummperOptions);
        list.forEach(element -> {
            subSequenceNode.getValue().add(new ScalarNode(Tag.STR, element, startMark, endMark,
                    DumperOptions.ScalarStyle.PLAIN));
        });

        mappingNode.getValue().add(new NodeTuple(
                new ScalarNode(Tag.STR, nodeName,
                        startMark, endMark, DumperOptions.ScalarStyle.PLAIN
                ), subSequenceNode));
    }
    public void Create(Datastructure datastructure, String toFile, Logger logger) {

        var subDataStructureSequenceYmlNode = new SequenceNode(
                Tag.SEQ, new ArrayList<>(), dummperOptions);

        datastructure.get_classname_list().forEach(className -> {

            var subDataStructureMappingYmlNode = new MappingNode(
                    Tag.MAP, new ArrayList<>(), dummperOptions);
            Datastructure.SubDataStructure subDataStructure = datastructure.get_datastructures_from_class_name(className);
            if (subDataStructure != null) {
                appendTupleNodeValueString(subDataStructureMappingYmlNode, "fqdn_class_name", subDataStructure.get_fqdn_class_name());
                appendTupleNodeValueString(subDataStructureMappingYmlNode, "filename", subDataStructure.get_filename());
                appendTupleNodeValueString(subDataStructureMappingYmlNode, "is_abstract", subDataStructure.is_abstract() ? "True" : "False");
                appendTupleNodeValueString(subDataStructureMappingYmlNode, "is_interface", subDataStructure.IsInterface() ? "True" : "False");
                appendTupleNodeValueString(subDataStructureMappingYmlNode, "is_inner_class", subDataStructure.isInnerClass() ? "True" : "False");
                appendTupleNodeValueString(subDataStructureMappingYmlNode, "filemodule", subDataStructure.get_filemodule());
                var fromImportsSequenceNode = new SequenceNode(
                        Tag.SEQ, new ArrayList<>(), dummperOptions);
                subDataStructure.get_from_imports().forEach((fqdnClassName, namespaceName) -> {
                    var fromImportsMappingNode = new MappingNode(
                            Tag.MAP, new ArrayList<>(), dummperOptions);
                    appendTupleNodeValueString(fromImportsMappingNode, "imported_class_name", fqdnClassName);
                    appendTupleNodeValueString(fromImportsMappingNode, "namespace_path", namespaceName);
                    fromImportsSequenceNode.getValue().add(fromImportsMappingNode);
                });
                appendTupleNodeValueSequenceNode(subDataStructureMappingYmlNode, "from_imports", fromImportsSequenceNode);

                appendSequence(subDataStructureMappingYmlNode,
                        subDataStructure.get_inner_class_name(),
                        "inner_classes");

                appendSequence(subDataStructureMappingYmlNode,
                        subDataStructure.getAnonymousInvocation(),
                        "anonymous_calls");

                appendSequence(subDataStructureMappingYmlNode,
                        subDataStructure.get_base_classes(),
                        "base_classes");

                var staticsSequenceNode = new SequenceNode(
                        Tag.SEQ, new ArrayList<>(), dummperOptions);
                subDataStructure.get_static_fields().forEach(staticType -> {
                    var staticsMappingNode = new MappingNode(
                            Tag.MAP, new ArrayList<>(), dummperOptions);
                    appendTupleNodeValueString(staticsMappingNode, "static_name", staticType.StaticName);
                    appendTupleNodeValueString(staticsMappingNode, "static_type", staticType.StaticType);
                    staticsSequenceNode.getValue().add(staticsMappingNode);
                });
                appendTupleNodeValueSequenceNode(subDataStructureMappingYmlNode, "statics", staticsSequenceNode);

                var variablesSequenceNode = new SequenceNode(
                        Tag.SEQ, new ArrayList<>(), dummperOptions);
                subDataStructure.get_variable_fields().forEach(variable -> {
                    var variablesMappingNode = new MappingNode(
                            Tag.MAP, new ArrayList<>(), dummperOptions);
                    appendTupleNodeValueString(variablesMappingNode, "variable_name", variable.variableName);
                    appendTupleNodeValueString(variablesMappingNode, "variable_type", variable.variableType);
                    appendTupleNodeValueString(variablesMappingNode, "is_member", variable.IsMember ? "True" : "False");
                    variablesSequenceNode.getValue().add(variablesMappingNode);
                });
                appendTupleNodeValueSequenceNode(subDataStructureMappingYmlNode, "variables", variablesSequenceNode);


                var methodsSequenceNode = new SequenceNode(
                        Tag.SEQ, new ArrayList<>(), dummperOptions);
                subDataStructure.get_method_fields().forEach(method -> {
                    var parameters = new SequenceNode(
                            Tag.SEQ, new ArrayList<>(), dummperOptions);
                    method.getParameters().forEach(parameter -> {
                        var methodParametersMappingNode = new MappingNode(
                                Tag.MAP, new ArrayList<>(), dummperOptions);
                        appendTupleNodeValueString(methodParametersMappingNode, "parameter_name", parameter.parameter);
                        appendTupleNodeValueString(methodParametersMappingNode, "parameter_type", parameter.userType);
                        parameters.getValue().add(methodParametersMappingNode);
                    });

                    var methodVariables = new SequenceNode(
                            Tag.SEQ, new ArrayList<>(), dummperOptions);
                    method.variables.forEach(variable -> {
                        var methodVariablesMappingNode = new MappingNode(
                                Tag.MAP, new ArrayList<>(), dummperOptions);
                        appendTupleNodeValueString(methodVariablesMappingNode, "variable_name", variable.variableName);
                        appendTupleNodeValueString(methodVariablesMappingNode, "variable_type", variable.variableType);
                        methodVariables.getValue().add(methodVariablesMappingNode);

                    });

                    var methodNode = new MappingNode(
                            Tag.MAP, new ArrayList<>(), dummperOptions);
                    appendTupleNodeValueString(methodNode, "method_name", method.methodName);
                    appendTupleNodeValueSequenceNode(methodNode, "parameters", parameters);
                    appendTupleNodeValueSequenceNode(methodNode, "method_variables", methodVariables);
                    appendTupleNodeValueString(methodNode, "is_private", method.IsPrivate ? "True" : "False");
                    methodsSequenceNode.getValue().add(methodNode);
                });
                appendTupleNodeValueSequenceNode(subDataStructureMappingYmlNode, "methods", methodsSequenceNode);

                var tmpMappingNode = new MappingNode( Tag.MAP, new ArrayList<>(), dummperOptions);
                tmpMappingNode.getValue().add(
                        new NodeTuple(
                                new ScalarNode(Tag.STR, "sub_datastructure",
                                        startMark, endMark, DumperOptions.ScalarStyle.PLAIN),
                                subDataStructureMappingYmlNode)
                );
                subDataStructureSequenceYmlNode.getValue().add(tmpMappingNode);
            } // if subDataStructure != null
        }); // for each classname

        Yaml yaml = new Yaml();
        try {
            logger.logInfo("Writing YAML file: " + toFile);
            PrintWriter writer = new PrintWriter(toFile);
            yaml.serialize(subDataStructureSequenceYmlNode, writer);
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }

    }
}